src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,167 @@
+package duke;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.IntStream;
+
+/**
+ * TaskList is the class to store all the tasks keyed in by the user during
+ * the session. It has a generic type of Task, which means the type of data
+ * stored in the list is Task. It is an immutable list and will always
+ * return a new list if modified.
+ *
+ * @author Muhammad Reyaaz
+ * @version %I% %G%
+ * @since 11
+ */
+class TaskList<T> {
+
+    private final ArrayList<Task> tasks;
+    //List starts numbering from 1
+    private final static int INCREMENT = 1;
+    /**
+     * Instantiate new list of tasks from an ArrayList
+     *
+     * @param newTasks Type TaskList<T>
+     */
+    TaskList(TaskList<T> newTasks) {
+        tasks = new ArrayList<Task>(newTasks.tasks);
+    }
+    /**
+     * Instantiate an empty ArrayList of tasks
+     */
+    protected TaskList() {
+        tasks = new ArrayList<Task>();
+    }
+    /**
+     * Only allow the TaskList class to instantiate the list of task from an
+     * ArrayList of previous tasks
+     */
+    private TaskList(ArrayList<Task> oldTasks) {
+        tasks = new ArrayList<Task>(oldTasks);
+    }
+    /**
+     * Get the total number of tasks in the list.
+     *
+     * @return sizeOfTasks
+     */
+    int numberOfTasks() {
+        return tasks.size();
+    }
+    /**
+     * Display to the user in a pretty format all the tasks in the list,
+     * starting from 1.
+     *
+     * @return TaskList<T>
+     *
+     */
+    TaskList<T> listAllTasks() {
+        AtomicLong counter = new AtomicLong();
+        tasks.stream().forEach(task ->
+                System.out.println(counter.incrementAndGet() + "." + task));
+        /*
+        for (int i=0; i < numberOfTasks(); i++) {
+            System.out.println(i + INCREMENT + "." + get(i));
+        }
+        */
+        return new TaskList<T>(tasks);
+    }
+
+    <T extends Task> TaskList<T> listFindTasks(String find, TaskList<T> list) {
+        AtomicLong counter = new AtomicLong();
+        tasks.stream().filter(task ->
+                task.getDescription().contains(find.substring(1))).forEach(task ->
+                        System.out.println(counter.incrementAndGet() + "." + task));
+        /*
+        for (int i=0; i < numberOfTasks(); i++) {
+            Task currTask = (Task)get(i);
+            if (currTask.getDescription().contains(find.substring(1))) {
+                System.out.println(i + INCREMENT + "." + currTask);
+            }
+        }
+        */
+        return new TaskList<T>(tasks);
+    }
+
+    /**
+     * Get a task in a specific position
+     * @return positionOfTask
+     */
+    Task get(int index) {
```

Comment:
getTask would be a better name  here
```suggestion
    Task getTask(int index) {
```